src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt

```diff
@@ -0,0 +1,66 @@
+package uk.co.llobile.kotlin.hours.types
+
+sealed class FunctionalOption<T>: Functor<T>, Applicative<T>, Monad<T> {
+
+    abstract fun isEmpty(): Boolean
+    abstract fun get(): T
+
+    override fun <R> map(fn: (T) -> R): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
+            Some(fn(get()))
+        }
+    }
+    override fun <R> apply(fn: Applicative<(T) -> R>): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
+            fn.map { it(get()) } as FunctionalOption<R>
```

Comment:
I suspect there's a declaration that doesn't require a cast here, but I couldn't get the generic types quite right.

Reply:
Also here use _when_ please.